testopt/testopt.go

```diff
@@ -0,0 +1,132 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package testopt provides objective functions for testing optimization algorithms.
+package testopt
+
+import (
+	"math"
+	"reflect"
+	"testing"
+
+	"github.com/gonum/floats"
+)
+
+// Minimum represents information about an optimal location of a function.
+type Minimum struct {
+	// X is the location of the minimum. If it is unknown or if the location is
+	// not a single point, X is nil.
+	X []float64
+	// F is the value of the objective function at X.
+	F float64
+	// Global indicates whether the minimum is local or global.
+	Global bool
+}
+
+// Function represents a test objective function.
+type Function interface {
+	F(x []float64) float64
+
+	// Minima returns known minima of the function. If no minima exist or if
+	// their location for the given dimension is unknown, nil is returned.
+	Minima() []Minimum
+}
+
+type Gradient interface {
+	Df(x, grad []float64)
+}
+
+type FunctionGradient interface {
+	FDf(x, grad []float64) float64
+}
+
+const (
+	defaultTol     = 1e-8
+	defaultGradTol = 1e-4
+)
+
+func testFunction(f Function, tests []funcTest, t *testing.T) {
```

Comment:
I've found fd.Gradient works surprisingly well,  even for function values
up in the 10^11 range. You could also specify an  [][2]float64 with  min/max
On Feb 4, 2015 4:09 PM, "Vladimír Chalupecký" <notifications@github.com>
wrote:

> In testopt/testopt.go
> <https://github.com/gonum/optimize/pull/63#discussion_r24132771>:
>
> > +}
> > +
> > +type Gradient interface {
> > +	Df(x, grad []float64)
> > +}
> > +
> > +type FunctionGradient interface {
> > +	FDf(x, grad []float64) float64
> > +}
> > +
> > +const (
> > +	defaultTol     = 1e-8
> > +	defaultGradTol = 1e-4
> > +)
> > +
> > +func testFunction(f Function, tests []funcTest, t *testing.T) {
>
> Using fd.Gradient() and some random locations has crossed my mind. The
> locations I test now are the standard starting locations and the optimal
> locations.
>
> I will try to add fd.Gradient() at random locations, but it might turn out
> as tricky to do in a general case. Some functions grow so fast away from a
> minimum that the numerical gradient can be very inaccurate. I'll see.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gonum/optimize/pull/63/files#r24132771>.
>